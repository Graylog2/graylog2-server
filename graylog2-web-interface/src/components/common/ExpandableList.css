:local(.list) {
    padding-left: 0;
    margin-bottom: 20px;
}

:local(.list) :local(.list) {
    margin-bottom: 0;
}
