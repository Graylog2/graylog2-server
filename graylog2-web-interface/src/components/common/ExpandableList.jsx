import React from 'react';
import PropTypes from 'prop-types';

import style from './ExpandableList.css';
/**
 * The ExpandableList will take a array or one of ExpandeableListItem to render
 * in list. This list can be expanded or flattened to give the user a overview
 * of categories. Inside the categories the user has the possibility of doing a selection.
 * The ExpandableList can be used nested.
 */
class ExpandableList extends React.Component {
  static propTypes = {
    /**
     * One or more elements of ExpandableListItem
     */
    children: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.arrayOf(PropTypes.element),
    ]),
  };

  static defaultProps = {
    children: [],
  };

  render() {
    const { children } = this.props;

    return (
      <ul className={style.list}>
        {children}
      </ul>
    );
  }
}

export default ExpandableList;
