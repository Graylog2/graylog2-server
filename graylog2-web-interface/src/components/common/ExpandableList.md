```js
const createReactClass = require('create-react-class');

const ExpandableListExample = createReactClass({

  render() {
    return (
      <ExpandableList>
        <ExpandableListItem selectable expanded header="Wheel of Time Character">
          <ExpandableList>
            <ExpandableListItem selectable header="Andor">
              <ExpandableList>
                <ExpandableListItem selectable expandable={false} header="Elayne" />
                <ExpandableListItem selectable expandable={false} header="Morgase" />
              </ExpandableList>
            </ExpandableListItem>
            <ExpandableListItem selectable header="Edmonds Field">
              <ExpandableList>
                <ExpandableListItem selectable expandable={false} header="Rand" />
                <ExpandableListItem selectable expandable={false} header="Perrin" />
                <ExpandableListItem selectable expandable={false} header="Egwene" />
                <ExpandableListItem checked expandable={false} header="Mat" />
                <ExpandableListItem expandable={false} selectable header="Nynaeve" />
              </ExpandableList>
            </ExpandableListItem>
          </ExpandableList>
        </ExpandableListItem>
      </ExpandableList>
    );
  },
});

<ExpandableListExample />

```
