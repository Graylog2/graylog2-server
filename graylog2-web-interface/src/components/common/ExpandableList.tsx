/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import type { PropsWithChildren } from 'react';
import React from 'react';
import PropTypes from 'prop-types';

import style from './ExpandableList.css';

type Props = PropsWithChildren<{
  className?: string,
}>;

/**
 * The ExpandableList will take a array or one of ExpandableListItem to render
 * in list. This list can be expanded or flattened to give the user a overview
 * of categories. Inside the categories the user has the possibility of doing a selection.
 * The ExpandableList can be used nested.
 */
const ExpandableList = ({ children, className }: Props) => {
  return (
    <ul className={className ? `${style.list} ${className}` : style.list}>
      {children}
    </ul>
  );
};

ExpandableList.defaultProps = {
  className: undefined,
};

ExpandableList.propTypes = {
  /**
   * One or more elements of ExpandableListItem
   */
  className: PropTypes.string,
};

export default ExpandableList;
