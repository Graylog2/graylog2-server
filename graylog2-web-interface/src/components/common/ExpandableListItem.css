:local(.listItemPadding) {
    padding: 10px 5px;
}

:local(.listItemContainer) {
    font-size: 14px;
    line-height: 20px;
}

:local(.listItemContainer) label {
    min-height: 20px;
    margin-bottom: 2px;
    margin-right: 5px;
}

:local(.listItemContainer) label * {
    cursor: pointer;
}

:local(.expandBoxContainer) {
    display: inline-block;
    width: 20px;
    margin-right: 5px;
}

:local(.expandBox).fa-stack {
    cursor: pointer;
    font-size: 1.4em;
    line-height: 20px;
    width: 1em;
    height: 1em;
    vertical-align: text-top;
}

:local(.expandBox):hover * {
    color: #731748;
}

:local(.iconBackground)  {
    color: #FFF;
}

:local(.header) {
    font-size: 14px;
}

:local(.subheader) {
    font-size: 0.95em;
    margin-left: 0.5em;
    color: #aaaaaa;
}

:local(.expandableContent) {
    border-left: 1px #eee solid;
    margin-left: 35px;
    margin-top: 10px;
    padding-left: 5px;
}