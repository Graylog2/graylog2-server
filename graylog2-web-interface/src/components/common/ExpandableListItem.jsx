import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from 'react-bootstrap';

import style from './ExpandableListItem.css';

/**
 * The ExpandableListItem is needed to render a ExpandableList.
 */
class ExpandableListItem extends React.Component {
  static propTypes = {
    /** Is the Item checked */
    checked: PropTypes.bool,
    /**
     * Indicates whether the checkbox on this item should be in an indetermined state or not.
     * This is mostly helpful to represent cases where the element is only partially checked,
     * for instance when ExpandableListItem's child is an ExpandableList and some of its items
     * are checked, but others are not.
     */
    indetermined: PropTypes.bool,
    /** Is the item selectable */
    selectable: PropTypes.bool,
    /** Is the Item expandable */
    expandable: PropTypes.bool,
    /** Is the Item expanded */
    expanded: PropTypes.bool,
    /** Forces to stay expanded regardless of clicking on the arrow */
    stayExpanded: PropTypes.bool,
    /** The header of the item */
    header: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
    /** The possible subheader of the item */
    subheader: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    /** Can be a html tag or again a ExpandableList */
    children: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.arrayOf(PropTypes.element),
    ]),
    /** Leave space before and after list item */
    padded: PropTypes.bool,
  };

  static defaultProps = {
    checked: false,
    indetermined: false,
    expandable: true,
    expanded: false,
    selectable: true,
    children: [],
    subheader: undefined,
    stayExpanded: false,
    padded: true,
  };

  state = {
    expanded: this.props.expanded,
    selectable: false,
    subheader: '',
  };

  componentDidMount() {
    if (this.props.indetermined && this._checkbox) {
      this._checkbox.indeterminate = this.props.indetermined;
    }
  }

  componentDidUpdate(prevProps) {
    if (prevProps.expanded !== this.props.expanded) {
      this._toggleExpand();
    }

    if (this._checkbox) {
      this._checkbox.indeterminate = this.props.indetermined;
    }
  }

  _checkbox = undefined;

  _toggleExpand = () => {
    if (this.props.stayExpanded) {
      this.setState({ expanded: true });
    } else {
      this.setState({ expanded: !this.state.expanded });
    }
  };

  _filterInputProps = (props) => {
    const { expanded, indetermined, stayExpanded, padded, ...inputProps } = props;
    return inputProps;
  };

  _clickOnHeader = () => {
    if (this._checkbox) {
      this._checkbox.click();
    }
  };

  render() {
    const { expanded } = this.state;
    const { checked, expandable, selectable, header, subheader, children, ...otherProps } = this.props;
    const headerToRender = selectable ? (<span role="button" tabIndex={0} onClick={this._clickOnHeader}>{header}</span>) : header;
    const inputProps = this._filterInputProps(otherProps);
    const liClassName = this.props.padded ? style.listItemPadding : '';

    return (
      <li className={liClassName}>
        <div className={style.listItemContainer}>
          {selectable && <Checkbox inputRef={(c) => { this._checkbox = c; }} inline checked={checked} {...inputProps} />}
          {expandable &&
          <div className={style.expandBoxContainer}>
            <div className={`fa-stack ${style.expandBox}`} role="button" tabIndex={0} onClick={this._toggleExpand}>
              <i className={`fa fa-circle-thin fa-stack-1x ${style.iconBackground}`} />
              <i className={`fa fa-stack-1x fa-angle-${expanded ? 'down' : 'up'}`} />
            </div>
          </div>
          }
          <span className={style.header}>{headerToRender}{subheader && <span className={style.subheader}>{subheader}</span>}</span>
        </div>
        <div className={style.expandableContent}>
          {expanded && children}
        </div>
      </li>
    );
  }
}

export default ExpandableListItem;
