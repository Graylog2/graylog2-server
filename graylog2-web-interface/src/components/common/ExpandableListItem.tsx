/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import { Checkbox } from 'components/bootstrap';

import Icon from './Icon';

const ItemWrap = styled.li(({ padded }: { padded: boolean }) => css`
  padding: ${padded ? '10px 5px' : ''};
`);

const Container = styled.div(({ theme }) => css`
  display: flex;
  font-size: ${theme.fonts.size.body};
  line-height: 20px;

  label {
    min-height: 20px;
    margin-bottom: 2px;
    margin-right: 5px;

    * {
      cursor: pointer;
    }
  }
`);

const Toggle = styled.div`
  display: inline-block;
  width: 20px;
  margin-right: 5px;
`;

const IconStack = styled.div(({ theme }) => css`
  &.fa-stack {
    cursor: pointer;
    font-size: ${theme.fonts.size.large};
    line-height: 20px;
    width: 1em;
    height: 1em;
    vertical-align: text-top;

    &:hover [class*='fa-'] {
      color: ${theme.colors.variant.primary};
      opacity: 1;
    }
  }

  [class*='fa-']:first-child {
    opacity: 0;

    ~ [class*='fa-']:hover {
      color: ${theme.colors.global.contentBackground};
    }
  }
`);

const HeaderWrap = styled.span(({ theme }) => css`
  font-size: ${theme.fonts.size.large};
`);

const Header = styled.button`
  display: flex;
  border: 0;
  padding: 0;
  text-align: left;
  background: transparent;
`;

const Subheader = styled.span(({ theme }) => css`
  font-size: ${theme.fonts.size.body};
  margin-left: 0.5em;
  color: ${theme.colors.gray[70]};
`);

const ExpandableContent = styled.div(({ theme }) => css`
  border-left: 1px ${theme.colors.gray[90]} solid;
  margin-left: 35px;
  margin-top: 10px;
  padding-left: 5px;
`);

const _filterInputProps = (props) => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { expanded, indetermined, stayExpanded, padded, ...inputProps } = props;

  return inputProps;
};

type Props = React.ComponentProps<typeof Checkbox> & {
  header: React.ReactNode,
  checked?: boolean,
  indetermined?: boolean,
  selectable?: boolean,
  expandable?: boolean,
  expanded?: boolean,
  stayExpanded?: boolean,
  subheader?: React.ReactNode,
  children?: React.ReactNode,
  padded?: boolean,
  readOnly?: boolean,
  onChange?: (e: React.MouseEvent<Checkbox>) => void,
};

type State = {
  expanded: boolean,
};

/**
 * The ExpandableListItem is needed to render a ExpandableList.
 */
class ExpandableListItem extends React.Component<Props, State> {
  private _checkbox: Checkbox | undefined;

  static propTypes = {
    /** Is the Item checked */
    checked: PropTypes.bool,
    /**
     * Indicates whether the checkbox on this item should be in an indetermined state or not.
     * This is mostly helpful to represent cases where the element is only partially checked,
     * for instance when ExpandableListItem's child is an ExpandableList and some of its items
     * are checked, but others are not.
     */
    indetermined: PropTypes.bool,
    /** Is the item selectable */
    selectable: PropTypes.bool,
    /** Is the Item expandable */
    expandable: PropTypes.bool,
    /** Is the Item expanded */
    expanded: PropTypes.bool,
    /** Forces to stay expanded regardless of clicking on the arrow */
    stayExpanded: PropTypes.bool,
    /** The header of the item */
    header: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
    /** The possible subheader of the item */
    subheader: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    /** Can be a html tag or again a ExpandableList */
    children: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.arrayOf(PropTypes.element),
    ]),
    /** Leave space before and after list item */
    padded: PropTypes.bool,
    /** Mark checkbox as read only */
    readOnly: PropTypes.bool,
    /** onChange handler for the checkbox */
    onChange: PropTypes.func,
  };

  static defaultProps = {
    checked: false,
    indetermined: false,
    expandable: true,
    expanded: false,
    selectable: true,
    children: [],
    subheader: undefined,
    stayExpanded: false,
    padded: true,
    readOnly: false,
    onChange: () => undefined,
  };

  constructor(props) {
    super(props);

    this.state = {
      expanded: props.expanded,
    };
  }

  componentDidMount() {
    const { indetermined } = this.props;

    if (indetermined && this._checkbox) {
      this._checkbox.indeterminate = indetermined;
    }
  }

  componentDidUpdate(prevProps) {
    const { expanded, indetermined } = this.props;

    if (prevProps.expanded !== expanded) {
      this._toggleExpand();
    }

    if (this._checkbox) {
      this._checkbox.indeterminate = indetermined;
    }
  }

  _toggleExpand = () => {
    const { stayExpanded } = this.props;
    const { expanded } = this.state;

    if (stayExpanded) {
      this.setState({ expanded: true });
    } else {
      this.setState({ expanded: !expanded });
    }
  };

  _clickOnHeader = () => {
    if (this._checkbox) {
      this._checkbox.click();
    }
  };

  render() {
    const { expanded } = this.state;
    const { padded } = this.props;
    const { checked, expandable, selectable, header, subheader, children, ...otherProps } = this.props;
    const headerToRender = selectable ? (<Header type="button" tabIndex={0} onClick={this._clickOnHeader}>{header}</Header>) : header;
    const inputProps = _filterInputProps(otherProps);

    return (
      <ItemWrap padded={padded}>
        <Container>
          {selectable && <Checkbox inputRef={(ref) => { this._checkbox = ref; }} inline checked={checked} {...inputProps} />}
          {expandable
          && (
            <Toggle>
              <IconStack className="fa-stack" tabIndex={0} onClick={this._toggleExpand}>
                <Icon name="circle" className="fa-stack-1x" />
                <Icon name={`angle-${expanded ? 'down' : 'up'}`} className="fa-stack-1x" />
              </IconStack>
            </Toggle>
          )}
          <HeaderWrap className="header">
            {headerToRender}
            {subheader && <Subheader>{subheader}</Subheader>}
          </HeaderWrap>
        </Container>

        <ExpandableContent>
          {expanded && children}
        </ExpandableContent>
      </ItemWrap>
    );
  }
}

export default ExpandableListItem;
