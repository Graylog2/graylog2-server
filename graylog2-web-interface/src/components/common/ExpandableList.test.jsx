import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import 'helpers/mocking/react-dom_mock';

import ExpandableList from 'components/common/ExpandableList';
import ExpandableListItem from 'components/common/ExpandableListItem';

describe('<ExpandableList />', () => {
  it('should render with no children', () => {
    const wrapper = renderer.create(<ExpandableList />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render with a Item', () => {
    const wrapper = renderer.create(<ExpandableList>
      <ExpandableListItem header="Wheel of time">
        <span>Edmonds Field</span>
      </ExpandableListItem>
    </ExpandableList>);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render with a nested ExpandableList', () => {
    const wrapper = renderer.create(<ExpandableList>
      <ExpandableListItem expandable expanded header="Wheel of time">
        <ExpandableList>
          <ExpandableListItem expandable expanded={false} header="Edmonds Field" />
        </ExpandableList>
      </ExpandableListItem>
    </ExpandableList>);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });


  it('should expand a expandable list item', () => {
    const wrapper = mount(<ExpandableList>
      <ExpandableListItem expandable header="Wheel of time" readOnly>
        <ExpandableList>
          <ExpandableListItem header="Edmonds Field" readOnly />
        </ExpandableList>
      </ExpandableListItem>
    </ExpandableList>);
    expect(wrapper.find('span.header').length).toBe(1);
    wrapper.find('div.fa-stack').simulate('click');
    expect(wrapper.find('span.header').length).toBe(2);
  });

  it('should select a selectable list item', () => {
    const checkFn = jest.fn();
    const wrapper = mount(<ExpandableList>
      <ExpandableListItem expanded header="Wheel of time" readOnly>
        <ExpandableList>
          <ExpandableListItem expanded selectable header="Edmonds Field" onChange={checkFn} />
        </ExpandableList>
      </ExpandableListItem>
    </ExpandableList>);
    wrapper.find('input[type="checkbox"]').at(1).simulate('change', { target: { checked: true } });
    expect(checkFn.mock.calls.length).toBe(1);
  });
});
