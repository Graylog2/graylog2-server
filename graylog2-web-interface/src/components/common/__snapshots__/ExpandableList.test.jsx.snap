// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpandableList /> should render with a Item 1`] = `
<ul
  className="list"
>
  <li
    className="listItemPadding"
  >
    <div
      className="listItemContainer"
    >
      <label
        className="checkbox-inline"
        title=""
      >
        <input
          checked={false}
          disabled={false}
          type="checkbox"
        />
      </label>
      <div
        className="expandBoxContainer"
      >
        <div
          className="fa-stack expandBox"
          onClick={[Function]}
          role="button"
          tabIndex={0}
        >
          <i
            className="fa fa-circle-thin fa-stack-1x iconBackground"
          />
          <i
            className="fa fa-stack-1x fa-angle-up"
          />
        </div>
      </div>
      <span
        className="header"
      >
        <span
          onClick={[Function]}
          role="button"
          tabIndex={0}
        >
          Wheel of time
        </span>
      </span>
    </div>
    <div
      className="expandableContent"
    />
  </li>
</ul>
`;

exports[`<ExpandableList /> should render with a nested ExpandableList 1`] = `
<ul
  className="list"
>
  <li
    className="listItemPadding"
  >
    <div
      className="listItemContainer"
    >
      <label
        className="checkbox-inline"
        title=""
      >
        <input
          checked={false}
          disabled={false}
          type="checkbox"
        />
      </label>
      <div
        className="expandBoxContainer"
      >
        <div
          className="fa-stack expandBox"
          onClick={[Function]}
          role="button"
          tabIndex={0}
        >
          <i
            className="fa fa-circle-thin fa-stack-1x iconBackground"
          />
          <i
            className="fa fa-stack-1x fa-angle-down"
          />
        </div>
      </div>
      <span
        className="header"
      >
        <span
          onClick={[Function]}
          role="button"
          tabIndex={0}
        >
          Wheel of time
        </span>
      </span>
    </div>
    <div
      className="expandableContent"
    >
      <ul
        className="list"
      >
        <li
          className="listItemPadding"
        >
          <div
            className="listItemContainer"
          >
            <label
              className="checkbox-inline"
              title=""
            >
              <input
                checked={false}
                disabled={false}
                type="checkbox"
              />
            </label>
            <div
              className="expandBoxContainer"
            >
              <div
                className="fa-stack expandBox"
                onClick={[Function]}
                role="button"
                tabIndex={0}
              >
                <i
                  className="fa fa-circle-thin fa-stack-1x iconBackground"
                />
                <i
                  className="fa fa-stack-1x fa-angle-up"
                />
              </div>
            </div>
            <span
              className="header"
            >
              <span
                onClick={[Function]}
                role="button"
                tabIndex={0}
              >
                Edmonds Field
              </span>
            </span>
          </div>
          <div
            className="expandableContent"
          />
        </li>
      </ul>
    </div>
  </li>
</ul>
`;

exports[`<ExpandableList /> should render with no children 1`] = `
<ul
  className="list"
/>
`;
